### ISSUE TITLE
Allow to compile cabal-install from source

### ISSUE THREAD

**Issue by jkachmar** [haskell/ghcup#68]
This was mentioned in #3, but it's of some interest to me as I'm adapting `ghcup` to work with Alpine at the moment and I'm compiling `cabal-install` from scratch.

It may still be overkill in the vast majority of use-cases though, so feel free to close if it's out of scope for this project.

**Comment by hasufell** [comment]
Can you give more detail?

I find it generally confusing to have this option since we are already basically compiling it from source in 2 steps:
1. download pre-built cabal-install binary
2. issue `cabal new-install cabal-install`

I suspect that doesn't work because of musl in alpine? In that case, it would make more sense to fix the missing `cabal-install` binary. @hvr

**Comment by jkachmar** [comment]
In this case, I’m not downloading the `cabal-install` binary at all. I’m downloading the `cabal-install` tarball from Hackage and installing it via the `bootstrap.sh` script in its root directory.

(I’m additionally overriding `bootstrap.sh`’s linker variable with a patch, but that is relatively specific to my use-case and probably out-of-scope for anything that `ghcup` should handle directly).

The ideal scenario would be if we could get `musl`-supporting GHC bindists at haskell.org, but the versions in aports itself are stale, so I’ve been taking a direct approach while I figure out what exactly is needed.

**Comment by hvr** [comment]
I did build http://downloads.haskell.org/cabal/cabal-install-2.4.1.0/cabal-install-2.4.1.0-x86_64-alpine-linux-musl.tar.xz for Alpine, does it not work anymore on current Alpines?

**Comment by hasufell** [comment]
indeed, it's also in download-urls: https://github.com/haskell/ghcup/blob/master/.download-urls#L71

**Comment by jkachmar** [comment]
Huh, I guess I missed this! I still think it would probably make sense to include a compilation function for cabal-install using the bootstrap script for completeness’ sake (e.g. for the truly paranoid who don’t want to trust external binaries). I could open a PR with my changes, if that might be more useful to discuss this, otherwise feel free to close this issue.

**Comment by hasufell** [comment]
> (e.g. for the truly paranoid who don’t want to trust external binaries)

Paranoid users should use their package manager, not shell scripts that download unverified tarballs (or binaries, doesn't matter).

ghcup is not there to replace a source distro PM or offer reproducible builds or anything like that. It's a lazy alternative. Security is hard.

**Comment by jkachmar** [comment]
Unfortunately I'm running into numerous linker errors when I try to execute:
```
apk add binutils-gold cabal gcc ghcl lvm5 wget
cabal update && cabal install cabal-install-2.4.1.0 -w ghc-8.4.3
```
...so it seems that this avenue is somewhat closed to me, whereas patching both the GHC source tree and the `cabal-install` bootstrapping process _appear_ to work without issue.

---

On the topic of having a `compile-cabal` option in general: `ghcup` isn't meant to cater to users who want to bootstrap GHC and its associated tooling, then why have the `compile` option for GHC at all?

If the tool provides one, it seems to make sense to provide the other, since `cabal-install` can be bootstrapped without a previous installation.

**Comment by hasufell** [comment]
> `ghcup` isn't meant to cater to users who want to bootstrap GHC and its associated tooling, then why have the `compile` option fo

That was a request from the author of the previous tool https://github.com/haskell/ghcup/issues/1

I never used this feature and I think most people don't. In addition, it is rather hard to support correctly across different GHC versions. It's almost like writing a package definition (PKGBUILD, APKBUILD, ebuild, ...).

The reason to support GHC compilation, imo, is more that you can mess with build-time config https://github.com/haskell/ghcup/blob/master/ghcup#L362
In order to build GHC, you already need GHC. So you're likely going to fetch a binary anyway. In order to compile cabal-install from source, you can just use `cabal new-install cabal-install`. That is why I don't really see the use case.

**Comment by hasufell** [comment]
I might be willing to have "unstable" commands (and hide them by default) in case other people maintain them. I think the `compile` subcommand should not show up in the default `--help` menu.
